### PR TITLE
liquid staking full implementation

### PR DESCRIPTION
**Pull request by 0xkobe**

### CODE REVIEW

**Review by zmanian — COMMENTED** [review]
Distribution/types/msg.go is still work in progress correct?

File: x/staking/types/expected_keepers.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ type AccountKeeper interface {
 
 	// TODO remove with genesis 2-phases refactor https://github.com/cosmos/cosmos-sdk/issues/2862
```
[the remark this answers is not in the document]
Comment (zmanian) [reply x/staking/types/expected_keepers.go:23]: I think so

File: x/staking/spec/02_state_transitions.md [hunk 2]
```diff
@@ -174,3 +174,36 @@ The total number of tokens is now `T + T_j`, and the total number of shares is `
...
+6. Delegate unbonded tokens from delegator address to the validator
+
+### Transfer tokenize share record
+
+The process of transferring the tokenize share record
+
+1. Check tokenize share record exists and the owner is the sender of the message
+2. Delete old owner's reference to tokenize share record
+3. Update tokenize share record to have new owner
+4. Add new owner's reference to tokenize share record
```
Comment (zmanian) [inline x/staking/spec/02_state_transitions.md:209]: Should mention that transferring the tokenized share record is done to move the reward withdrawal rights

File: x/staking/types/msg.go [hunk 3]
```diff
@@ -24,6 +24,7 @@ var (
 	_ sdk.Msg                            = &MsgDelegate{}
 	_ sdk.Msg                            = &MsgUndelegate{}
 	_ sdk.Msg                            = &MsgBeginRedelegate{}
+	_ sdk.Msg                            = &MsgTransferTokenizeShareRecord{}
```
Comment (0xkobe) [inline x/staking/types/msg.go:27]: Missing msg: MsgTokenizeShares, MsgRedeemTokensforShares

File: x/staking/types/codec.go [hunk 4]
```diff
@@ -17,6 +17,7 @@ func RegisterLegacyAminoCodec(cdc *codec.LegacyAmino) {
 	cdc.RegisterConcrete(&MsgDelegate{}, "cosmos-sdk/MsgDelegate", nil)
 	cdc.RegisterConcrete(&MsgUndelegate{}, "cosmos-sdk/MsgUndelegate", nil)
 	cdc.RegisterConcrete(&MsgBeginRedelegate{}, "cosmos-sdk/MsgBeginRedelegate", nil)
+	cdc.RegisterConcrete(&MsgTransferTokenizeShareRecord{}, "cosmos-sdk/MsgTransferTokenizeShareRecord", nil)
```
Comment (0xkobe) [inline x/staking/types/codec.go:20]: Missing msg: MsgTokenizeShares, MsgRedeemTokensforShares

File: x/staking/types/codec.go [hunk 5]
```diff
@@ -27,6 +28,7 @@ func RegisterInterfaces(registry types.InterfaceRegistry) {
 		&MsgDelegate{},
 		&MsgUndelegate{},
 		&MsgBeginRedelegate{},
+		&MsgTransferTokenizeShareRecord{},
```
Comment (0xkobe) [inline x/staking/types/codec.go:31]: Missing msg: MsgTokenizeShares, MsgRedeemTokensforShares

**Comment by 0xkobe** [comment]
Moved to non-epoch solution, Hence closing this PR.

**Pull request closed without merging**